warehouse/static/js/warehouse/controllers/password_breach_controller.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Controller } from "stimulus";
+import { debounce } from "debounce";
+
+export default class extends Controller {
+  static targets = ["password", "message"];
+
+  connect() {
+    this.check = debounce(this.check, 1000);
+    this._lastCheckedPassword = null;
+  }
+
+  check() {
+    if (this.passwordTarget.value !== this._lastCheckedPassword) {
+      this.hideMessage();
+      if (this.passwordTarget.value.length > 2) {
+        this._lastCheckedPassword = this.passwordTarget.value;
+        return this.checkPassword(this.passwordTarget.value).catch(
+          e => {
+            console.error(e);  // eslint-disable-line no-console
+            this.hideMessage();  // default to hiding the message on errors
+          }
+        );
+      }
+    }
+    return null;
+  }
+
+  async checkPassword(password) {
+    let digest = await this.digestMessage(password);
+    let hex = this.hexString(digest);
+    let response = await fetch(this.getURL(hex));
+    if (response.ok === false) {
+      const msg = "Error while validating hashed password, disregard on development";
+      console.error(`${msg}: ${response.status} ${response.statusText}`);  // eslint-disable-line no-console
+    } else {
+      let text = await response.text();
+      this.parseResponse(text, hex);
+    }
+  }
+
+  async digestMessage(message) {
+    // https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest
+    const encoder = new TextEncoder();
+    const data = encoder.encode(message);
+    return window.crypto.subtle.digest("SHA-1", data);
```

Comment:
According to [the spec](https://github.com/w3c/webcrypto/issues/28) `window.crypto.subtle` is `undefined` in insecure  contexts, and as a result I get a `TypeError` on this line when trying to run it locally over HTTP.

I'm curious a) how you were able to work around this when testing and b) how you suggest we handle this for local development.

Reply:
That's news to me 😅

The capture shows the controller working locally and I just tested it successfully using `http://localhost/account/register/` in Chrome, Firefox and Safari on my Mac.

How exactly are you getting the `TypeError`?

Reply:
Huh, weird. I use a custom host mapping to 127.0.0.1, and `window.crypto.subtle` is `undefined` when I use that host, but not when I use `localhost`/`127.0.0.1`, even though they're all HTTP.

Guess I'll just have to ignore this! 😅